File: datasets/Part 1 - Data Preprocessing/Section 2 -------------------- Part 1 - Data Preprocessing --------------------/Ejercicios/Plantilla_Preprocesado_de_Datos.py
# -*- coding: utf-8 -*-
"""
Created on Tue Dec 10 15:21:50 2019

@author: Aguilerimon
"""

import numpy as np
import matplotlib.pyplot as ptl
import pandas as pd

#Importar el data set
dataset = pd.read_csv('Data.csv')

#Separamos las variables del data set

#Localizacion de los elementos por localizacion (index)

#Asignamos a la variable x todas las filas de las tres primeras columnas como 
#variables independientes del data set
x = dataset.iloc[:,:-1].values

#Asignamos a la variable y todas las filas de la ultima columna como variable dependiente
y = dataset.iloc[:, 3].values

#Tratamiento de los NAs o valores faltantes
from sklearn.preprocessing import Imputer

#Especificamos todos los terminos necesarios para el tratamiento de los NAs
imputer = Imputer(missing_values = "NaN", strategy = "mean", axis = 0)

imputer = imputer.fit(x[:, 1:3])
x[:, 1:3] = imputer.transform(x[:, 1:3])

#Codificar los datos categóricos
#Import de las librerias LabelEncoder(Le da igual si los datos son catogórico u ordinales)
#Y la libreria OneHotEncoder(Implementación de las variable dummy categóricas)
from sklearn.preprocessing import LabelEncoder, OneHotEncoder

#Constructor del método labelencoder
labelencoder_x = LabelEncoder()
labelencoder_y = LabelEncoder()

#Aplicamos la transformación numérica a la primer columna y la sobreescribimos al objeto x
x[:,0] = labelencoder_x.fit_transform(x[:,0])

#Para los valores boolean no nos interesa utilizar el método OneHotEncoder 
y = labelencoder_y.fit_transform(y)

#Es necesario haber tratado anteriormente el vector con LabelEncoder
#debido a que OneHotEncoder no podrá convertir el string original a float
onehotencoder = OneHotEncoder(categorical_features=[0])
x = onehotencoder.fit_transform(x).toarray()

#Dividir el dataset en un conjunto de entrenamiento y de evaluación
#Importamos la sublibreria train_test_split
from sklearn.cross_validation import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x,y, test_size= 0.2, random_state=0)

#Escalado de variables
from sklearn.preprocessing import StandardScaler
sc_x = StandardScaler()
x_train = sc_x.fit_transform(x_train)
x_test = sc_x.transform(x_test)

File: datasets/Part 1 - Data Preprocessing/Section 2 -------------------- Part 1 - Data Preprocessing --------------------/Ejercicios/Plantilla_Preprocesado_de_Datos.R
# Plantilla para el pre Procesado de Datos
# Importar el dataset
dataset = read.csv('Data.csv')

#Tratamiento de los valores NA
dataset$Age = ifelse(is.na(dataset$Age),
                     ave(dataset$Age, FUN = function(x) mean(x,na.rm = TRUE)), 
                     dataset$Age)

dataset$Salary = ifelse(is.na(dataset$Salary),
                     ave(dataset$Salary, FUN = function(x) mean(x,na.rm = TRUE)), 
                     dataset$Salary)

#Codificar variables categóricas
dataset$Country = factor(dataset$Country,
                         levels = c("France", "Spain", "Germany"),
                         labels = c(1, 2, 3))

dataset$Purchased = factor(dataset$Purchased,
                           levels = c("No", "Yes"),
                           labels = c(0, 1))

#Dividir los datos en conjunto de entrenamiento y evaluación
#Importamos la libreria para separar los conjuntos
library(caTools)

#Especificamos el numero de repeticiones del algoritmo (Puede ser cualquiera)
set.seed(123)

#Devuelve un vector del mismo tamaño que la variable o vector que se le da
#indicando si debe formar parte del conjunto de entrenamiento
split = sample.split(dataset$Purchased, SplitRatio = 0.8)

#Separa y almacena en dos diferentes variables los conjuntos de 
#entrenamiento y evaluación
training_set = subset(dataset, split==TRUE)
testing_set = subset(dataset, split==FALSE)

#Escalado de variables
training_set[,2:3] = scale(training_set[,2:3])
testing_set[,2:3] = scale(testing_set[,2:3])